Clean up illink suppressions

sbomer opened the pull request:
The warnings suppressed in https://github.com/dotnet/runtime/pull/96327 were not observed in dotnet/runtime because:
- When trimming the library individually, IL2065 was suppressed, and
- When trimming the OOB, the code was unreachable

The per-library trimming phase still had a bunch of analysis warnings suppressed that were never removed as part of https://github.com/dotnet/runtime/issues/38033. This removes the NoWarn suppressions, and fixes up the ILLink.Suppressions.LibraryBuild.xml logic so that these xml suppressions get used during the library build. Since corelib has two separate suppression files (one shared, and one for coreclr), this changes the logic to allow multiple suppression xml files.

sbomer committed:
Clean up illink suppressions

sbomer committed:
Suppress IL2008

sbomer committed:
Support multiple suppression xmls

CoreLib has a shared ILLink.Substitutions.LibraryBuild.xml, and
one that is specific to coreclr. Use an ItemGroup to support the use
of both.

sbomer committed:
Suppress IL2035

ViktorHofer commented on eng/illink.targets:
```diff
@@ -210,23 +213,15 @@
       <!-- pass the non-embedded descriptors xml file on the command line -->
       <ILLinkArgs Condition="'$(ILLinkDescriptorsLibraryBuildXml)' != ''">$(ILLinkArgs) -x "$(ILLinkDescriptorsLibraryBuildXml)"</ILLinkArgs>
       <ILLinkArgs Condition="'$(ILLinkSubstitutionsLibraryBuildXml)' != ''">$(ILLinkArgs) --substitutions "$(ILLinkSubstitutionsLibraryBuildXml)"</ILLinkArgs>
+      <ILLinkArgs Condition="'@(ILLinkSuppressionsLibraryBuildXml)' != ''">$(ILLinkArgs) --link-attributes &quot;@(ILLinkSuppressionsLibraryBuildXml->'%(FullPath)', '&quot; --link-attributes &quot;')&quot;</ILLinkArgs>
```
You shouldn't need the outer quotes here:

```suggestion
      <ILLinkArgs Condition="'@(ILLinkSuppressionsLibraryBuildXml)' != ''">$(ILLinkArgs) --link-attributes "@(ILLinkSuppressionsLibraryBuildXml->'%(FullPath)', '&quot; --link-attributes &quot;')"</ILLinkArgs>
```

I'm not even sure about the inner quotes. `&quot;` is usually only required in item includes: `<Item Include="&quot;...&quot;" />`

ViktorHofer commented on eng/illink.targets:
```diff
@@ -210,23 +213,15 @@
-      <LinkerNoWarn>$(LinkerNoWarn);IL2062;IL2063;IL2064;IL2065;IL2066</LinkerNoWarn>
-      <!-- IL2067-IL2091: Unsatisfied DynamicallyAccessedMembers requirements -->
-      <LinkerNoWarn>$(LinkerNoWarn);IL2067;IL2068;IL2069;IL2070;IL2071;IL2072;IL2073;IL2074;IL2075;IL2076;IL2077;IL2078;IL2079;IL2080;IL2081;IL2082;IL2083;IL2084;IL2085;IL2086;IL2087;IL2088;IL2089;IL2090;IL2091</LinkerNoWarn>
+      <LinkerNoWarn>$(LinkerNoWarn);IL2008;IL2009;IL2035</LinkerNoWarn>
+      <LinkerNoWarn>$(LinkerNoWarn);IL2025</LinkerNoWarn>
```
```suggestion
      <LinkerNoWarn>$(LinkerNoWarn);IL2008;IL2009;IL2025;IL2035</LinkerNoWarn>
```

ViktorHofer commented on eng/illink.targets:
```diff
@@ -210,23 +213,15 @@
-      <LinkerNoWarn>$(LinkerNoWarn);IL2067;IL2068;IL2069;IL2070;IL2071;IL2072;IL2073;IL2074;IL2075;IL2076;IL2077;IL2078;IL2079;IL2080;IL2081;IL2082;IL2083;IL2084;IL2085;IL2086;IL2087;IL2088;IL2089;IL2090;IL2091</LinkerNoWarn>
+      <LinkerNoWarn>$(LinkerNoWarn);IL2008;IL2009;IL2035</LinkerNoWarn>
+      <LinkerNoWarn>$(LinkerNoWarn);IL2025</LinkerNoWarn>
       <!-- IL2121: Unnecessary UnconditionalSuppressMessage attribute -->
       <LinkerNoWarn>$(LinkerNoWarn);IL2121</LinkerNoWarn>
```
Please also move this one up into L223 and the comment next to the others

eerhardt replied:
Is suppressing IL2121 still necessary? Where are the unnecessary UnconditionalSuppressMessage attributes? And why are they still needed?

ViktorHofer commented on eng/illink.targets:
```diff
@@ -27,27 +27,30 @@
     <ILLinkResourcesSubstitutionIntermediatePath>$(IntermediateOutputPath)ILLink.Resources.Substitutions.xml</ILLinkResourcesSubstitutionIntermediatePath>
     <GenerateResourcesSubstitutions Condition="'$(GenerateResourcesSubstitutions)' == '' and '$(StringResourcesPath)' != ''">true</GenerateResourcesSubstitutions>
+ 
+    <ILLinkSuppressionsXmlPrefix>$(ILLinkDirectory)ILLink.Suppressions</ILLinkSuppressionsXmlPrefix>
+    <ILLinkSuppressionsXml>$(ILLinkSuppressionsXmlPrefix).xml</ILLinkSuppressionsXml>
+    <ILLinkSuppressionsConfigurationSpecificXml>$(ILLinkSuppressionsXmlPrefix).$(Configuration).xml</ILLinkSuppressionsConfigurationSpecificXml>
```
Why did you move those down? I would prefer to keep them where they were to keep the diff smaller. It's currently hard to understand what actually changed. Including the `EnableTrimAnalyzer` property below.

ViktorHofer commented on eng/illink.targets:
```diff
@@ -27,27 +27,30 @@
+    <ILLinkSuppressionsXml>$(ILLinkSuppressionsXmlPrefix).xml</ILLinkSuppressionsXml>
+    <ILLinkSuppressionsConfigurationSpecificXml>$(ILLinkSuppressionsXmlPrefix).$(Configuration).xml</ILLinkSuppressionsConfigurationSpecificXml>
+  </PropertyGroup>
+  <ItemGroup>
```
Please keep the empty lines between property and item groups as that makes reading this file easier (at least for me).

sbomer committed:
Replace unnecessary &quot;

sbomer committed:
PR feedback

- Keep properties in existing order
- Separate PropertyGroup/ItemGroup with blank lines
- Place all NoWarns on one line

sbomer committed:
Update eng/illink.targets

Co-authored-by: Viktor Hofer <viktor.hofer@microsoft.com>
